### DISCUSSION TITLE
HTTP downloads get stuck in EFI mode

### DISCUSSION THREAD

**Discussion by korekhov** [ipxe/ipxe#1242] [General]
Hello, all!

Last week I ran into something puzzling, to say the least, and I don't really know how to proceed from iPXE side to troubleshooting this issue.

The problem statement - the same baremetal machines on "slower" international links (100G w/ ~148ms round trip times) are getting stuck while iPXE in EFI mode is trying to download kernel and/or initrd files. The strangest thing is that if I update boot method to legacy one, the same machine would download the same files w/o any issues and proceed with PXEboot.

I've took the latest iPXE sources, recompiled with "DEBUG=httpcore:3,http" and my simple embedded script (i can post it if needed), but the only messages I get are these:

```
>>Checking Media Presence......
>>Media Present......
>>Start PXE over IPv4 on MAC: <mac>
  Station IP address is <ip>

  Server IP address is <server-ip>
  NBP filename is ipxe.efi
  NBP filesize is 1113600 Bytes

>>Checking Media Presence......
>>Media Present......
 Downloading NBP file...

  NBP file downloaded successfully.
iPXE initialising devices...
autoexec.ipxe... Not found (https://ipxe.org/2d12618e)

iPXE 1.21.1+ (g821b) -- Open Source Network Boot Firmware -- https://ipxe.org
Features: DNS HTTP HTTPS iSCSI TFTP VLAN SRP AoE EFI Menu
Configuring (net0 <mac>)........ ok
pxelinux.cfg/01-<mac>.ipxe... ok
HTTP 0x5b94fbc8 http://<company>.com:8080/httpboot/pxelinux.cfg/overrides.ipxe
http://<company>.com:8080/httpboot/pxelinux.cfg/overrides.ipxe...
HTTP 0x5b94fbc8 TX GET /httpboot/pxelinux.cfg/overrides.ipxe HTTP/1.1
HTTP 0x5b94fbc8 TX Connection: keep-alive
HTTP 0x5b94fbc8 TX User-Agent: iPXE/1.21.1+ (g821b)
HTTP 0x5b94fbc8 TX Host: <company>.com:8080
HTTP 0x5b94fbc8 RX HTTP/1.1 200 OK
HTTP 0x5b94fbc8 RX Content-Type: application/octet-stream
HTTP 0x5b94fbc8 RX Last-Modified: Tue, 02 Apr 2024 22:29:35 GMT
HTTP 0x5b94fbc8 RX Content-Length: 3085
HTTP 0x5b94fbc8 RX X-Content-Type-Options: nosniff
HTTP 0x5b94fbc8 RX Server: foreman-proxy/3.6.2
HTTP 0x5b94fbc8 RX Date: Tue, 25 Jun 2024 23:02:53 GMT
HTTP 0x5b94fbc8 RX Connection: Keep-Alive
 ok
HTTP 0x5b9509c8 http://<company>.com:8080/httpboot/vmlinuz
http://<company>.com:8080/httpboot/vmlinuz...
HTTP 0x5b9509c8 TX GET /httpboot/vmlinuz HTTP/1.1
HTTP 0x5b9509c8 TX Connection: keep-alive
HTTP 0x5b9509c8 TX User-Agent: iPXE/1.21.1+ (g821b)
HTTP 0x5b9509c8 TX Host: <company>.com:8080
HTTP 0x5b9509c8 RX HTTP/1.1 200 OK
HTTP 0x5b9509c8 RX Content-Type: application/octet-stream
HTTP 0x5b9509c8 RX Last-Modified: Thu, 16 May 2024 22:22:44 GMT
HTTP 0x5b9509c8 RX Content-Length: 11686952
HTTP 0x5b9509c8 RX X-Content-Type-Options: nosniff
HTTP 0x5b9509c8 RX Server: foreman-proxy/3.6.2
HTTP 0x5b9509c8 RX Date: Tue, 25 Jun 2024 23:02:53 GMT
HTTP 0x5b9509c8 RX Connection: Keep-Alive
 ok
HTTP 0x5b9528a8 http://<company>.com:8080/httpboot/initrd.lz
http://<company>.com:8080/httpboot/initrd.lz...
HTTP 0x5b9528a8 TX GET /httpboot/initrd.lz HTTP/1.1
HTTP 0x5b9528a8 TX Connection: keep-alive
HTTP 0x5b9528a8 TX User-Agent: iPXE/1.21.1+ (g821b)
HTTP 0x5b9528a8 TX Host: <company>.com:8080
HTTP 0x5b9528a8 RX HTTP/1.1 200 OK
HTTP 0x5b9528a8 RX Content-Type: application/octet-stream
HTTP 0x5b9528a8 RX Last-Modified: Fri, 24 May 2024 21:02:33 GMT
HTTP 0x5b9528a8 RX Content-Length: 47922242
HTTP 0x5b9528a8 RX X-Content-Type-Options: nosniff
HTTP 0x5b9528a8 RX Server: foreman-proxy/3.6.2
HTTP 0x5b9528a8 RX Date: Tue, 25 Jun 2024 23:03:07 GMT
HTTP 0x5b9528a8 RX Connection: Keep-Alive
 15%
```

At this point in time, this issue happens only in one particular location, which according to networking team is exactly the same as the others, just a bit slower. Unfortunately, I currently do not have any EFI-capable systems in other locations to experiment more and find out of that's truly a location-specific problem or not. The only other EFI booted systems we have are in "faster" locations and do not exhibit this issue when HTTP just never completes.

Has anyone else experienced such issues? Any advices/suggestions on how to proceed and/or get more debugging information at this point? I'm working on getting some packet capturing set up upstream of my systems, but was wondering if there are more DEBUG flags I could use, maybe...

Thanks!

**Comment by NiKiZe** [comment]
What is the output of ifstat (in iPXE) im looking for which driver is in use.

Yes there is reports of other slow or stuck downloads, driver/nic or firmware bugs mostly. See the search function on GitHub, try terms such as "slow" and "stuck" and check issues and discussions. #1170 might be similar, but let's not jump to conclusions.

You only have this efi system? Other hardware where it works is also different?

**Reply by korekhov** [reply]
```
<snip>
 Operation canceled (https://ipxe.org/0b072095)
iPXE> ifstat
net0: <mac> using 14e4-16D7 on 0000:65:00.0 (Ethernet) [open]
  [Link:up, TX:27828 TXE:38 RX:40611 RXE:0]
  [TXE: 1 x "Network unreachable (https://ipxe.org/28086090)"]
  [TXE: 37 x "Error 0x2a2c4089 (https://ipxe.org/2a2c4089)"]
net1: <mac> using 14e4-16D7 on 0000:65:00.1 (Ethernet) [closed]
  [Link:down, TX:0 TXE:0 RX:0 RXE:0]
  [Link status: Unknown (https://ipxe.org/1a086194)]
net2: <mac> using NII on NII-0000:00:14.0 (Ethernet) [closed]
  [Link:down, TX:0 TXE:0 RX:0 RXE:0]
  [Link status: Unknown (https://ipxe.org/1a086194)]

iPXE> ping <gw>
0 bytes from <none>: seq=1: Connection timed out (https://ipxe.org/4c1b2092)
0 bytes from <none>: seq=2: Connection timed out (https://ipxe.org/4c1b2092)
0 bytes from <none>: seq=3: Connection timed out (https://ipxe.org/4c1b2092)
Finished: Operation canceled (https://ipxe.org/0b072095)
iPXE>
```

But network engineer says no error counters increased on corresponding switch port.

These are 25G Broadcoms, BTW:
```
# lspci | grep -i eth
65:00.0 Ethernet controller: Broadcom Inc. and subsidiaries BCM57414 NetXtreme-E 10Gb/25Gb RDMA Ethernet Controller (rev 01)
65:00.1 Ethernet controller: Broadcom Inc. and subsidiaries BCM57414 NetXtreme-E 10Gb/25Gb RDMA Ethernet Controller (rev 01)
# ethtool -i eth0
driver: bnxt_en
version: 5.15.0-107-generic
firmware-version: 218.0.219.13/pkg 218.0.219.21
<snip>
```

And just to clarify, this is not just one system that exhibits this issue, I have many like that in that particular location. I also have the same kind of systems that are "closer" in network hops and they are not affected by this issue even if booted in EFI mode.

Just to reiterate, my remote and troubled systems do NOT have this problem when booted in legacy mode, which basically eliminates possible switch/cable/network issues/misconfiguration. Unfortunately, switching these systems into legacy boot mode permanently is not an option.

I'm rebuilding ipxe.efi with "DEBUG=netdevice,bnxt" at the moment, so will share the results in a bit.

Thanks!

**Comment by korekhov** [comment]
DEBUG=bnxt,netdevice was giving a ton of below messages making it hard not to miss other messages, but both kernel and initrd were fully downloaded. Trying without "bnxt" now.

```
assert(!iob) failed at drivers/net/bnxt/bnxt.c line 527
assert(!iob) failed at drivers/net/bnxt/bnxt.c line 527
assert(!iob) failed at drivers/net/bnxt/bnxt.c line 527
assert(!iob) failed at drivers/net/bnxt/bnxt.c line 527
assert(!iob) failed at drivers/net/bnxt/bnxt.c line 527
assert(!iob) failed at drivers/net/bnxt/bnxt.c line 527
assert(!iob) failed at drivers/net/bnxt/bnxt.c line 527
assert(!iob) failed at drivers/net/bnxt/bnxt.c line 527
assert(!iob) failed at drivers/net/bnxt/bnxt.c line 527
assert(!iob) failed at drivers/net/bnxt/bnxt.c line 527
assert(!iob) failed at drivers/net/bnxt/bnxt.c line 527
assert(!iob) failed at drivers/net/bnxt/bnxt.c line 527
assert(!iob) failed at drivers/net/bnxt/bnxt.c line 527
assert(!iob) failed at drivers/net/bnxt/bnxt.c line 527
```

**Comment by korekhov** [comment]
Without "bnxt", I get this:

```

>>Checking Media Presence......
>>Media Present......
>>Start PXE over IPv4 on MAC: <net0-mac>
  Station IP address is <net0-ip>

  Server IP address is <server-ip>
  NBP filename is ipxe.efi
  NBP filesize is 1126400 Bytes

>>Checking Media Presence......
>>Media Present......
 Downloading NBP file...

  NBP file downloaded successfully.
iPXE initialising devices...
NETDEV net0 registered (phys MNP-0000:65:00.0 hwaddr <net0-mac>)
NETDEV net0 transmission 0x5b945490 failed: Network unreachable (https://ipxe.org/28086090)
NETDEV net0 link is up
NETDEV net0 opening
autoexec.ipxe... Not found (https://ipxe.org/2d12618e)
NETDEV net0 closing
NETDEV net0 unregistered
NETDEV net0 registered (phys 0000:65:00.0 hwaddr <net0-mac>)
NETDEV net0 transmission 0x5b944890 failed: Network unreachable (https://ipxe.org/28086090)
NETDEV net1 registered (phys 0000:65:00.1 hwaddr <net1-mac>)
NETDEV net2 registered (phys NII-0000:00:14.0 hwaddr <net2-mac>)


iPXE 1.21.1+ (g77acf) -- Open Source Network Boot Firmware -- https://ipxe.org
Features: DNS HTTP HTTPS iSCSI TFTP VLAN SRP AoE EFI Menu
NETDEV net0 opening
NETDEV net0 link is up
NETDEV net0 starting configuration via ipv6
NETDEV net0 starting configuration via dhcp
Configuring (net0 <net0-mac>)...NETDEV net0 configured via ipv6
...NETDEV net0 configured via dhcp
 ok
pxelinux.cfg/01-<net0-mac>.ipxe... ok
http://<company>.com:8080/httpboot/pxelinux.cfg/overrides.ipxe... ok
http://<company>.com:8080/httpboot/vmlinuz...NETDEV net0 transmission 0x5b954790 failed: Error 0x2a2c4089 (https://ipxe.org/2a2c4089)
NETDEV net0 transmission 0x5b954790 failed: Error 0x2a2c4089 (https://ipxe.org/2a2c4089)
NETDEV net0 transmission 0x5b954790 failed: Error 0x2a2c4089 (https://ipxe.org/2a2c4089)
NETDEV net0 transmission 0x5b954790 failed: Error 0x2a2c4089 (https://ipxe.org/2a2c4089)
NETDEV net0 transmission 0x5b954790 failed: Error 0x2a2c4089 (https://ipxe.org/2a2c4089)
NETDEV net0 transmission 0x5b954790 failed: Error 0x2a2c4089 (https://ipxe.org/2a2c4089)
NETDEV net0 transmission 0x5b954790 failed: Error 0x2a2c4089 (https://ipxe.org/2a2c4089)
NETDEV net0 transmission 0x5b954790 failed: Error 0x2a2c4089 (https://ipxe.org/2a2c4089)
NETDEV net0 transmission 0x5b954790 failed: Error 0x2a2c4089 (https://ipxe.org/2a2c4089)
NETDEV net0 transmission 0x5b954790 failed: Error 0x2a2c4089 (https://ipxe.org/2a2c4089)
NETDEV net0 transmission 0x5b954790 failed: Error 0x2a2c4089 (https://ipxe.org/2a2c4089)
NETDEV net0 transmission 0x5b954790 failed: Error 0x2a2c4089 (https://ipxe.org/2a2c4089)
NETDEV net0 transmission 0x5b954790 failed: Error 0x2a2c4089 (https://ipxe.org/2a2c4089)
NETDEV net0 transmission 0x5b954790 failed: Error 0x2a2c4089 (https://ipxe.org/2a2c4089)
NETDEV net0 transmission 0x5b954790 failed: Error 0x2a2c4089 (https://ipxe.org/2a2c4089)
NETDEV net0 transmission 0x5b954790 failed: Error 0x2a2c4089 (https://ipxe.org/2a2c4089)
  9%NETDEV net0 transmission 0x5b954790 failed: Error 0x2a2c4089 (https://ipxe.org/2a2c4089)
  9%NETDEV net0 transmission 0x5b954790 failed: Error 0x2a2c4089 (https://ipxe.org/2a2c4089)
  9%NETDEV net0 transmission 0x5b954790 failed: Error 0x2a2c4089 (https://ipxe.org/2a2c4089)
  9%NETDEV net0 transmission 0x5b954790 failed: Error 0x2a2c4089 (https://ipxe.org/2a2c4089)
  9%NETDEV net0 transmission 0x5b954790 failed: Error 0x2a2c4089 (https://ipxe.org/2a2c4089)
 Operation canceled (https://ipxe.org/0b072095)
iPXE> NETDEV net0 transmission 0x5b94bf90 failed: Error 0x2a2c4089 (https://ipxe.org/2a2c4089)
NETDEV net0 transmission 0x5b94bf90 failed: Error 0x2a2c4089 (https://ipxe.org/2a2c4089)

iPXE> ifstat
NETDEV net0 transmission 0x5b94bf90 failed: Error 0x2a2c4089 (https://ipxe.org/2a2c4089)

net0: <net0-mac> using 14e4-16D7 on 0000:65:00.0 (Ethernet) [open]
  [Link:up, TX:553 TXE:25 RX:992 RXE:0]
  [TXE: 1 x "Network unreachable (https://ipxe.org/28086090)"]
  [TXE: 24 x "Error 0x2a2c4089 (https://ipxe.org/2a2c4089)"]
net1: <net1-mac> using 14e4-16D7 on 0000:65:00.1 (Ethernet) [closed]
  [Link:down, TX:0 TXE:0 RX:0 RXE:0]
  [Link status: Unknown (https://ipxe.org/1a086194)]
net2: <net2-mac> using NII on NII-0000:00:14.0 (Ethernet) [closed]
  [Link:down, TX:0 TXE:0 RX:0 RXE:0]
  [Link status: Unknown (https://ipxe.org/1a086194)]
iPXE> NETDEV net0 transmission 0x5b94bf90 failed: Error 0x2a2c4089 (https://ipxe.org/2a2c4089)
NETDEV net0 transmission 0x5b94bf90 failed: Error 0x2a2c4089 (https://ipxe.org/2a2c4089)
```

As you can see it stuck again. Error's link is suggesting to enable "bnxt" in my DEBUG, but I already done that and that did not provide any useful (to me) info other than "asset" error messages I posted above.

What "Error: No buffer space available" indicates? And why this does not happen in legacy boot mode on the same machine?

Is there anything else I could try from iPXE side of things?

Thanks!

**Comment by NiKiZe** [comment]
Which binary are you using for legacy boot?
You seem to be using ipxe.efi, so for legacy the "same" would be ipxe.pxe.

However I would like you to compare normal boot between ipxe.efi, and snp.efi to see if this is driver related.

**Reply by korekhov** [reply]
undionly.kpxe is used for legacy boots.

Let me try snp.efi.

**Reply by korekhov** [reply]
so, snp.efi seemed to be able to handle this better as it did not get stuck, but kept retrying until it fully loaded my kernel and initrd even though it counted a ton of TX errors:

```
<snip>
NETDEV net0 transmission 0x0 failed: Error 0x2a086089 (https://ipxe.org/2a086089)
NETDEV net0 transmission 0x0 failed: Error 0x2a086089 (https://ipxe.org/2a086089)
NETDEV net0 transmission 0x0 failed: Error 0x2a086089 (https://ipxe.org/2a086089)
NETDEV net0 transmission 0x0 failed: Error 0x2a086089 (https://ipxe.org/2a086089)
NETDEV net0 transmission 0x0 failed: Error 0x2a086089 (https://ipxe.org/2a086089)
NETDEV net0 transmission 0x0 failed: Error 0x2a086089 (https://ipxe.org/2a086089)
NETDEV net0 transmission 0x0 failed: Error 0x2a086089 (https://ipxe.org/2a086089)
NETDEV net0 transmission 0x0 failed: Error 0x2a086089 (https://ipxe.org/2a086089)
 ok
net0: <net0-mac> using NII on NII-0000:65:00.0 (Ethernet) [open]
  [Link:up, TX:28891 TXE:29223 RX:58258 RXE:0]
  [TXE: 1 x "Network unreachable (https://ipxe.org/28086090)"]
  [TXE: 29222 x "Error 0x2a086089 (https://ipxe.org/2a086089)"]
net1: <net1-mac> using NII on NII-0000:65:00.1 (Ethernet) [closed]
  [Link:down, TX:0 TXE:0 RX:0 RXE:0]
  [Link status: Unknown (https://ipxe.org/1a086194)]
```

Should I consider this an optimal outcome and switch over to snp.efi? This would still leave a question of why ipxe.efi not able to work the same in "slower" location while it works in all other locations on the same HW...

**Reply by NiKiZe** [reply]
Could you verify if legacy boot with ipxe.pxe has issues in this location, also verify if it is without issues in the "good" location.

If behaviour differs, can you verify if bios/firmware versions differs between the machines?

**Reply by korekhov** [reply]
So, machines in question in both "good" and "bad" locations are exactly same in terms HW, their BIOS and NIC FW.

In "bad" location, only ipxe.efi (when booting in default EFI mode) exhibits the stuck-http issue. If I change to legacy boot mode with undionly.kpxe on the same "bad" machine, HTTP downloads work perfectly fine.

In "good" location, neither EFI (w/ ipxe.efi) or legacy (w/ undionly.kpxe) gave me any grief.

Both undionly.kpxe and ipxe.efi binaries are the same in all locations as well.

I'm testing snp.efi in all of my locations now as according to all of the docs and posts I've read so far, snp.efi came a long and seems like it should work just fine on all of our UEFI systems as they are 3-4 years old at the most.

But if you have any more suggestions or ideas to try with ipxe.efi, I'm all ears.

Thanks a lot!

**Reply by NiKiZe** [reply]
Still want to know how ipxe.pxe behaves.

Your tests so far suggests that only the ipxe native driver has issues, a test with ipxe.pxe could confirm if that is the case.

It would also be interesting to see if the change in the issue I linked is related to this or not.

**Reply by korekhov** [reply]
Somehow ipxe.pxe is even worse than ipxe.efi as former does not print any debug messages nor proceeding any further than this:

```
Broadcom UNDI PXE-2.1 v218.0.219.1
Copyright (C) 2000-2020 Broadcom Limited
Copyright (C) 1997-2000 Intel Corporation
All rights reserved.

CLIENT MAC ADDR: <mac>  GUID: 29E87C00-92B2-11EE-8000-84160C0B80C6  [09;00HDHCP./ 
CLIENT IP: <net0-ip>  MASK: <netmask>  DHCP IP: <dhcp-server-ip>
GATEWAY IP: <gw-ip>
PXE->EB: !PXE at 94E3:0040, entry point at 94E3:00D6
         UNDI code segment 94E3:4F20, data segment 896A:B790 (549-616kB)
         UNDI device is PCI 65:00.0, type DIX+802.3
         616kB free base memory after PXE unload
iPXE initialising devices...

iPXE 1.21.1+ (g77acf) -- Open Source Network Boot Firmware -- https://ipxe.org
Features: DNS HTTP HTTPS iSCSI TFTP SRP VLAN AoE ELF MBOOT PXE bzImage Menu PXEXT
Press Ctrl-B for command line
```

As per my embedded script, next step here is supposed to be "dhcp" command, but that does not happen. Even if I drop to command prompt, "dhcp" does not work for me either. My undionly.kpxe still works the same way it did before though...

Unless you have some other ideas, I'm going to try this now:
```
 *    e) Intercontinental WAN: expected bandwidth 5MB/s, typical RTT
 *       250ms, minimum required window 1280kB.
```
Thanks!

**Reply by korekhov** [reply]
Nope, 1280kB max tcp window size did not solve ipxe.efi problem for me and slightly slowed down both undionly.kpxe and snp.efi.

HEAP_SIZE=8MB is the next one...

**Reply by korekhov** [reply]
That did not work either, but setting TCP_MAX_WINDOW_SIZE=256k did and ipxe.efi started working for me.
It is approximately twice slower than snp.efi.

Also, the size of ipxe.efi is 3 times more than snp.efi, so what would be an advantage of using ipxe.efi for me in this situation then?
That snp.efi *may* not work on all systems?

Anyway, @NiKiZe - thank you so much for all your help, I really appreciate it!
Off to my next step, which is HTTPS with company-issued CA and it's certs. My first attempt was only 50% successful, so I may start yet another discussion after browsing through existing ones.

**Reply by korekhov** [reply]
BTW, if there's still you wanted me to try in regards to my ipxe.efi problem in a "bad" location, please let me know.
Otherwise, we can close this discussion, I guess.

**Comment by ptempier** [comment]
Getting hit by this issue i guess,
but not in efi mode or its not reported to me so i dont  investigate, as it's random.

Even on vmware in bios mode kpxe 32b binary.
The http download get stuck any random percentage, it happen ~15% of the time.

i tried using kkpxe, reducing windows tcp size, increasing the heap, also the broadcom fix of less buffers.
i always think it works, until i reboot for the Nth time and then it get stuck.

**Reply by NiKiZe** [reply]
If you don't have broadcom, then this is not the same.

You should try to run git bisect if you can find a older version that works.

You should not use kkpxe, either use ipxe.pxe or undionly.kpxe for pcbios pxe boots.
